playbooks/openstack/inventory.py

```diff
@@ -16,52 +16,74 @@
+    inventory['masters'] = {'hosts': []}
+    inventory['etcd'] = {'hosts': []}
+    inventory['infra_hosts'] = {'hosts': []}
+    inventory['app'] = {'hosts': []}
+    inventory['glusterfs'] = {'hosts': []}
+    inventory['dns'] = {'hosts': []}
+    inventory['lb'] = {'hosts': []}
+    inventory['localhost'] = {'ansible_connection': 'local'}
+
+    for server in cluster_hosts:
+        if 'group' in server.metadata:
+            group = server.metadata.get('group')
+            if group not in inventory:
+                inventory[group] = {'hosts': []}
+            inventory[group]['hosts'].append(server.name)
 
-    masters = [server.name for server in cluster_hosts
-               if server.metadata['host-type'] == 'master']
+        inventory['cluster_hosts']['hosts'].append(server.name)
 
-    etcd = [server.name for server in cluster_hosts
-            if server.metadata['host-type'] == 'etcd']
-    if not etcd:
-        etcd = masters
+        if server.metadata['host-type'] == 'master':
+            inventory['masters']['hosts'].append(server.name)
+            nodes.add(server.name)
+            osev3.add(server.name)
 
-    infra_hosts = [server.name for server in cluster_hosts
-                   if server.metadata['host-type'] == 'node' and
-                   server.metadata['sub-host-type'] == 'infra']
+        if server.metadata['host-type'] == 'etcd':
+            inventory['etcd']['hosts'].append(server.name)
+            osev3.add(server.name)
 
-    app = [server.name for server in cluster_hosts
-           if server.metadata['host-type'] == 'node' and
-           server.metadata['sub-host-type'] == 'app']
+        if (server.metadata['host-type'] == 'node' and
+                server.metadata['sub-host-type'] == 'infra'):
+            inventory['infra_hosts']['hosts'].append(server.name)
+            nodes.add(server.name)
+            osev3.add(server.name)
 
-    cns = [server.name for server in cluster_hosts
-           if server.metadata['host-type'] == 'cns']
+        if (server.metadata['host-type'] == 'node' and
+                server.metadata['sub-host-type'] == 'app'):
+            inventory['app']['hosts'].append(server.name)
+            nodes.add(server.name)
+            osev3.add(server.name)
 
-    nodes = list(set(masters + infra_hosts + app + cns))
+        if server.metadata['host-type'] == 'cns':
+            inventory['cns']['hosts'].append(server.name)
+            nodes.add(server.name)
+            osev3.add(server.name)
 
-    dns = [server.name for server in cluster_hosts
-           if server.metadata['host-type'] == 'dns']
+        if server.metadata['host-type'] == 'dns':
+            inventory['dns']['hosts'].append(server.name)
 
-    load_balancers = [server.name for server in cluster_hosts
-                      if server.metadata['host-type'] == 'lb']
+        if server.metadata['host-type'] == 'lb':
+            inventory['load_balancers']['hosts'].append(server.name)
```

Comment:
The inventory group key should be "lb" here to remain consistent with the previous code.